Simple drag and drop

Markvanderuit opened the pull request:
We now have simple blocks that are draggable and stretchable in all directions, in a very large
scrollable area that can also handle resizes. Next up we'll put them in a grid of some sort.

Commit:
Added basic drag-and-drop and button events

Markvanderuit committed:
- Ported drag functionality into a pane, so we can add content to it.
- Made main view scrollable, so it can expand (up to settable limits)
- NOT WORKING: started adding resize functionality to panes.

Markvanderuit committed:
- Made TimetableBlock draggable AND stretchable.
- Made TimetableBlock work in a large scrollable area.
- Fixed some checkstyle errors.

Markvanderuit committed:
Merge branch 'develop' into simple-drag-and-drop

bart-de-jonge commented on src/main/java/gui/TimetableBlock.java:
```diff
@@ -0,0 +1,198 @@
+        setLayoutY(100.0);
+        setMinHeight(100.0);
+        setMinWidth(100.0);
+        setStyle(normalStyle);
```
Could these settings be tweakable? Like make them fields with getters and setters, maybe even initialize them in the constructor

martijnjanssen replied:
I believe these are just settings for testing, not final values. Am I correct @Ezlo123 ?

bart-de-jonge replied:
But in the final versions, x, y, width and height values must be settable, therefore i suggest to put them in fields. The testing values of 100.0 can remain for now.

Markvanderuit replied:
Just testing settings. Did, however, add 2 constructors that allow you to specifiy coordinates and width, if necessary.

martijnjanssen commented on src/main/java/gui/RootCenterArea.java:
```diff
@@ -1,13 +1,43 @@
+        contentPane.getChildren().add(rect1);
+        contentPane.getChildren().add(rect2);
+        contentPane.getChildren().add(rect3);
+        contentPane.getChildren().add(rect4);
```
Is there no `addAll()` method that can be used here?

Markvanderuit replied:
Sure, but these are just for testing. Seems unnecessary when they'll be gone tomorrow.

bart-de-jonge commented on src/main/java/gui/TimetableBlock.java:
```diff
@@ -0,0 +1,198 @@
+     * @param pane pane this block is located in.
+     * @param event the mousedrag event in question.
+     */
+    private void onMouseDraggedHelper(RootCenterArea pane, MouseEvent event) {
```
You could also return an eventhandler with the body of this method as the body of the handle method here. That would clean up the code in the constructor

Markvanderuit replied:
Could you clarify this a bit more?

bart-de-jonge replied:
In the constructor you now make new eventhandler and call this method in the body, why not return an actual eventhandler from this function and add it like: `setOnMouseReleased(onMouseDraggedHelper())` seems a lot cleaner to me

martijn9612 commented:
Right now, when dragging the bottom edge of a `TimetableBlock` the edge stops at the upper bound. When doing this to the top edge, the entire block slides. Is this fixable?

Could we also specify a minimum height for the element?

martijn9612 commented:
The mouse also changes in position relative to the block when exiting the screen when dragging the block. Can we remove this? It would be better to have the mouse reset to the initial position, right?

Ezlo123 commented:
@martijn9612 I'm fixing the first dragging issue right now. However, the position change relative to the block (when your mouse exits the screen) would probably mean forcing the mouse to move. That's one of the things you should, from a user's point of view, probably never do. Not to mention that I honestly don't know if the OS actually lets me do that.

martijn9612 commented:
We'll discuss this later, I can't really explain what I mean over git. We are meeting again soon!

Markvanderuit committed:
Implemented most suggested improvements.
- Added extra constructors to block class for ease of use.
- Fixed an issue where blocks could be dragged against their will one way, but not the other.
- Fixed a really annoying rounding error causing blocks to shrink when dragged against an edge with force.

Markvanderuit committed:
Merge branch 'develop' into simple-drag-and-drop

bartdejonge1996 commented:
Great work fixing those bugs
